Fix a case where shipping cost is always free

henkkelder opened the pull request:
On a system with dutch NLS the calcultations on line 3250 and 3260 simply did not work. The sum of the calulations was always 0. This caused the cart to always show 'free shipping'.
Also, since only one amount is returned there is not much point in calculating both with as without tax.

<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Solves an issue where the shipping was always free. For some reason the sum of shipping remained 0.
| Type?         | bug fix 
| Category?     | FO
| BC breaks?    | Does it break backward compatibility? no
| Deprecations? | Does it deprecate an existing feature? no
| Fixed ticket? | (optional) If this PR fixes an [Issue](https://github.com/PrestaShop/PrestaShop/issues), please write "Fixes #[issue number]" here.
| How to test?  | Please indicate how to best verify that this PR is correct.

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/12710)
<!-- Reviewable:end -->

henkkelder committed:
Update Cart.php

On a system with dutch NLS the calcultations on line 3250 and 3260 simply did not work. The sum of the calulations was always 0. 
This caused the cart to always show 'free shipping'.
Also, since only one amount is returned there is not much point in calculating both with as without tax.

Quetzacoalt91 commented:
Hi @henkkelder,

We're unable to see how the suggested changes may help you in the described issue, as there is no clear changes in the method behavior, only some refacto.
Do you have a concrete example which could help us?

henkkelder commented:
I totally agree. There appears nothing wrong with the original code. But for some reason beyond my understanding (40 years of experience as a software developer) the flawless code still did not work on the Prestashop installation of my daughter webshop (http://www.faithandletters.com).
The array containing the sum of the shipping components remained 0. And yes, I did another installation where the calculation worked fine.

To be honest I consider the original code as of doubtable quality due to the complex way it does the calculation of both the "with", as "without tax" amounts, where only one summation is required.
So I modified the code and bingo, my daughters site suddenly showed shipping costs.

Maybe its because the installation uses dutch language settings. I can only guess. I did notice that appearantly more users suffer from shipping costs not calculated. I read a lot of guesses for the cause of that problem.

In my effort to find the problem I added debug statements to show both the amounts being summed as the result. For me it was clear that the calculation itself went wrong. So I rewrote it. Not the first time a rewrote solved a problem.

Below my logging. 
The lines with > contain the amounts to be summed;
The lines with -> show the sommation result within the loop
The lines with => show the sommation below the loop.
As you can see the sums remain 0.

> with tax: 7.37, referer: http://www.faithandletters.com/nl/bestelling
> without tax: 7.37, referer: http://www.faithandletters.com/nl/bestelling

-> with tax: 0, referer: http://www.faithandletters.com/nl/bestelling
-> without tax: 0, referer: http://www.faithandletters.com/nl/bestelling

=> with tax: 0, referer: http://www.faithandletters.com/nl/bestelling
=> without tax: 0, referer: http://www.faithandletters.com/nl/bestelling

Quetzacoalt91 commented on classes/Cart.php:
```diff
@@ -3233,27 +3233,29 @@ public function getTotalShippingCost($delivery_option = null, $use_tax = true, C
         }
-
-        return ($use_tax) ? $_total_shipping['with_tax'] : $_total_shipping['without_tax'];
+        return $_total_shipping;
```
An additional line is required to make the lint job pass.

```suggestion

        return $_total_shipping;
```

Quetzacoalt91 commented on classes/Cart.php:
```diff
@@ -3233,27 +3233,29 @@ public function getTotalShippingCost($delivery_option = null, $use_tax = true, C
             $default_country = new Country(Context::getContext()->cookie->id_country);
         }
-        if (null === $delivery_option) {
+        if (is_null($delivery_option)) {
```
We worked on the whole project to have a comparison instead of `is_null`. Please bring back the old version.

```suggestion
        if (null === $delivery_option) {
```

Quetzacoalt91 committed:
Update classes/Cart.php

Co-Authored-By: henkkelder <henkkelder@users.noreply.github.com>

Quetzacoalt91 committed:
Update classes/Cart.php

Co-Authored-By: henkkelder <henkkelder@users.noreply.github.com>

Commit:
add blank line

MathiasReker commented on classes/Cart.php:
```diff
@@ -3287,6 +3290,7 @@ public function getCarrierCost($id_carrier, $useTax = true, Country $default_cou
             }
         }
 
+        
```
Why an extra line?

henkkelder replied:
removed it

MathiasReker commented on classes/Cart.php:
```diff
@@ -3237,23 +3237,26 @@ public function getTotalShippingCost($delivery_option = null, $use_tax = true, C
+        return $_total_shipping;
     }
 
+
```
Why an extra line?

MathiasReker reviewed (commented):
Some extra lines was added

Commit:
Revert "add blank line"

This reverts commit e762f84b91a3899a2acc9f9a5a5684452ca8aec7.

remove 1 blank line that was wrongfully added.

Quetzacoalt91 commented:
Indeed, please have a look at the Travis builds in order to know what must be changed.

henkkelder commented:
Well I gonna leave it with this. I tried to solve a specific bug I ran into. I do not want to spend a lot of time trying to understand how your build system works. The log from the Travis build I see (detail in the 'build failed' line) doesn't tell me what the problem is. I go back to my own domain.

Quetzacoalt91 commented:
As our role is to make contributing easy, reading this kind of message means that we fail our objective. Especially when we modified several time this task to show what the issue is (https://travis-ci.org/PrestaShop/PrestaShop/jobs/503275251#L648-L667), then how to fix it (https://travis-ci.org/PrestaShop/PrestaShop/jobs/503275251#L671-L674).
We must dig into this.

LimousinArcherie commented:
Just to say that the fixing file works for me too. (PS 1.7.5.1)

matks commented:
> Hi @henkkelder,
> 
> We're unable to see how the suggested changes may help you in the described issue, as there is no clear changes in the method behavior, only some refacto.
> Do you have a concrete example which could help us?

Only change that can have an impact is it uses floats instead of ints, so maybe in a weird case the `+=` was not working (maybe when `$delivery_option_list[$id_address][$key]['total_price_with_tax']` has a weird type)

henkkelder commented:
In my situation the old code didnt work, my modification made it work. I also have no explanation for it. But the fact remains that the old code did not work.




I also considered the old code doing useless and unneccesary things like calculating both with as without tax totals.




Were I come from ther are situations were code doesnt act as expected and a refactor sometimes helps. I take the pragmatic approach there. 








Outlook voor Android downloaden







On Wed, Apr 10, 2019 at 4:21 PM +0200, "Mathieu Ferment" <notifications@github.com> wrote:













Hi @henkkelder,


We're unable to see how the suggested changes may help you in the described issue, as there is no clear changes in the method behavior, only some refacto.

Do you have a concrete example which could help us?



Only change that can have an impact is it uses floats instead of ints, so maybe in a weird case the += was not working (maybe when $delivery_option_list[$id_address][$key]['total_price_with_tax'] has a weird type)



—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

MathiasReker commented:
I rebased the PR and fixed lint issues: https://github.com/PrestaShop/PrestaShop/pull/14626

sarahdib commented:
@henkkelder 

Thank you for your contribution.

Can you list the steps to reproduce ?
I didn't reproduce the shipping free

marionf commented:
Hello @henkkelder 

Since we have not had answer for more than 30 days, I close this PR.
Please before submitting a PR, open an issue to describe the issue and how to reproduce it.
